java-checks/src/main/java/org/sonar/java/checks/BatchSQLStatementsCheck.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2024 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.EnumSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.MethodMatchers;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.LambdaExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import static org.sonar.java.checks.helpers.TreeHelper.findClosestParentOfKind;
+
+/**
+ *
+ */
+@Rule(key = "S6912")
+public class BatchSQLStatementsCheck extends IssuableSubscriptionVisitor {
+  private static final String MESSAGE = "Use \"addBatch\" and \"executeBatch\" to execute multiple SQL statements in a single call.";
+  private static final MethodMatchers EXECUTE_METHODS = MethodMatchers.create()
+    .ofSubTypes("java.sql.Statement")
+    .names("execute", "executeQuery", "executeUpdate")
+    .withAnyParameters()
+    .build();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.METHOD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    var methodTree = (MethodTree) tree;
+
+    var invocationsCollector = new StatementExecuteInvocationsCollector();
+    methodTree.accept(invocationsCollector);
+
+    var inLambdaInvocationsCollector = new InLambdaStatementExecuteInvocationsCollector();
+    methodTree.accept(inLambdaInvocationsCollector);
+
+    if (invocationsCollector.shouldReport() || inLambdaInvocationsCollector.shouldReport()) {
+      invocationsCollector.invocations().forEach(invocation -> reportIssue(invocation, MESSAGE));
+      inLambdaInvocationsCollector.invocations().forEach(invocation -> reportIssue(invocation, MESSAGE));
+    }
```

Comment:
These seem like two disjoint pieces of logic to me: 1 for the `invocationsCollector`, one for the `inLambdaInvocationsCollector`. I think it would be a little more readable to have them in two separate if-statements.

Reply:
This logic is a bit problematic because there are scenarios that won't be covered:

- when there is a single invocation not in loop in both, neither of them will be reported (with current logic and separating the two if)
- when there is a single invocation not in loop in one and multiple invocations on the other, then the current logic will correctly report both

Reply:
Changed the implementation to report only on loops